
## Tutorial on How to Handle Missing Data:

Real-world data often has missing values. For example imagine you have a dataset of students (boys & girls) in a college. Unfortunately, on previewing the data you find few students height and weight are missing. Now let us ask some questions to ourselves like

  - Can you accurately find a way out off handling the missing heights/ weights?
  - Should one just pretend as if the missing instances isn’t missing.?
  - Should one go and ask those students for their height/ weight? (Not feasible!)
  - Or can one just predict the shape of the missing instances based on previous experience?

Data can have missing values for a number of reasons such as: observations that were not recorded and data corruption. Handling missing data is important as many machine learning algorithms do not support data with missing values.

In this tutorial, you will learn how to handle missing data for machine learning with Python.

Specifically, after goinf through this tutorial you will know:

    How to marking invalid or corrupt values as missing in your dataset.
    How to remove rows with missing data from your dataset.
    How to impute missing values with mean values in your dataset.

Let’s get started.



### Overview

The Pima Indians Diabetes Dataset involves predicting the onset of diabetes within 5 years in Pima Indians given medical details.

It is a binary (2-class) classification problem. The number of observations for each class is not balanced. There are 768 observations with 8 input variables and 1 output variable. The variable names are as follows:

0. Number of times pregnant.
1. Plasma glucose concentration a 2 hours in an oral glucose tolerance test.
2. Diastolic blood pressure (mm Hg).
3. Triceps skinfold thickness (mm).
4. 2-Hour serum insulin (mu U/ml).
5. Body mass index (weight in kg/(height in m)^2).
6. Diabetes pedigree function.
7. Age (years).
8. Class variable (0 or 1).

We can load the dataset as a Pandas DataFrame and print summary statistics on each attribute.

In [1]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import os
df = pd.read_csv('diabetes.csv',header = None)

## 1. Preview the Data

In [2]:
df.describe()

0    1    2    3    4    5      6    7    8
count   769  769  769  769  769  769    769  769  769
unique   18  137   48   52  187  249    518   53    3
top       1   99   70    0    0   32  0.258   22    0
freq    135   17   57  227  374   13      6   72  500

This dataset is known to have missing values. Specifically, there are missing observations for some columns that are marked as a zero value. This is useful. We can see that there are columns that have a minimum value of zero (0). On some columns, a value of zero does not make sense and indicates an invalid or missing value.

Specifically, the following columns have an invalid zero minimum value: Plasma glucose concentration, Diastolic blood pressure, Triceps skinfold thickness, 2-Hour serum insulin, Body mass index

Let’ confirm this my looking at the raw data, the example prints the first 20 rows of data.

In [3]:
df.head(20)

0        1              2              3        4     5  \
0   Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI   
1             6      148             72             35        0  33.6   
2             1       85             66             29        0  26.6   
3             8      183             64              0        0  23.3   
4             1       89             66             23       94  28.1   
5             0      137             40             35      168  43.1   
6             5      116             74              0        0  25.6   
7             3       78             50             32       88    31   
8            10      115              0              0        0  35.3   
9             2      197             70             45      543  30.5   
10            8      125             96              0        0     0   
11            4      110             92              0        0  37.6   
12           10      168             74              0        0    38   
13           10      139             80              0        0  27.1   
14            1      189             60             23      846  30.1   
15            5      166             72             19      175  25.8   
16            7      100              0              0        0    30   
17            0      118             84             47      230  45.8   
18            7      107             74              0        0  29.6   
19            1      103             30             38       83  43.3   

                           6    7        8  
0   DiabetesPedigreeFunction  Age  Outcome  
1                      0.627   50        1  
2                      0.351   31        0  
3                      0.672   32        1  
4                      0.167   21        0  
5                      2.288   33        1  
6                      0.201   30        0  
7                      0.248   26        1  
8                      0.134   29        0  
9                      0.158   53        1  
10                     0.232   54        1  
11                     0.191   30        0  
12                     0.537   34        1  
13                     1.441   57        0  
14                     0.398   59        1  
15                     0.587   51        1  
16                     0.484   32        1  
17                     0.551   31        1  
18                     0.254   31        1  
19                     0.183   33        0

We can get a count of the number of missing values on each of these columns. We can do this my marking all of the values in the subset of the DataFrame we are interested in that have zero values as True. We can then count the number of true values in each column.

We can do this my marking all of the values in the subset of the DataFrame we are interested in that have zero values as True. We can then count the number of true values in each column.

## 2. Check Missing Values

In [4]:
print((df[[1,2,3,4,5]] == '0').sum())

1      5
2     35
3    227
4    374
5     11
dtype: int64


We can see that columns 1,2 and 5 have just a few zero values, whereas columns 3 and 4 show a lot more, nearly half of the rows.

This highlights that different “missing value” strategies may be needed for different columns, e.g. to ensure that there are still a sufficient number of records left to train a predictive model.

## 3. Marking Missing Values

In Python, specifically Pandas, NumPy and Scikit-Learn, we mark missing values as NaN. Values with a NaN value are ignored from operations like sum, count, etc. We can do this by using the replace() function on a subset of the columns we are interested in.

In [5]:
# mark zero values as missing or NaN
df[[1,2,3,4,5]] = df[[1,2,3,4,5]].replace('0', np.NaN)

# count the number of NaN values in each column
print(df.isnull().sum())

0      0
1      5
2     35
3    227
4    374
5     11
6      0
7      0
8      0
dtype: int64


In [6]:
print(df.head(20))

              0        1              2              3        4     5  \
0   Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI   
1             6      148             72             35      NaN  33.6   
2             1       85             66             29      NaN  26.6   
3             8      183             64            NaN      NaN  23.3   
4             1       89             66             23       94  28.1   
5             0      137             40             35      168  43.1   
6             5      116             74            NaN      NaN  25.6   
7             3       78             50             32       88    31   
8            10      115            NaN            NaN      NaN  35.3   
9             2      197             70             45      543  30.5   
10            8      125             96            NaN      NaN   NaN   
11            4      110             92            NaN      NaN  37.6   
12           10      168             74            

You can see we have succesfully marked the missing values using replace function.



## 4. Remove Rows With Missing Values

The simplest strategy for handling missing data is to remove records that contain a missing value. We can do this by creating a new Pandas DataFrame with the rows containing missing values removed.

Pandas provides the dropna() function that can be used to drop either columns or rows with missing data.

In [7]:
# drop rows with missing values
df.dropna(inplace=True)

# summarize the number of rows and columns in the dataset
print(df.shape)

(393, 9)


Running this example, we can see that the number of rows has been aggressively cut from 768 in the original dataset to 393 with all rows containing a NaN removed.

## 5. Impute Missing Values

Imputing refers to using a model to replace missing values. There are many options we could consider when replacing a missing value, for example:

- A constant value that has meaning within the domain, such as 0, distinct from all other values.
- A value from another randomly selected record.
- A mean, median or mode value for the column.
- A value estimated by another predictive model.

Pandas provides the fillna() function for replacing missing values with a specific value.

In [8]:
# fill missing values with mean column values
df.fillna(df.mean(), inplace=True)

# count the number of NaN values in each column
print(df.isnull().sum())

0    0
1    0
2    0
3    0
4    0
5    0
6    0
7    0
8    0
dtype: int64


Running the example shows that all NaN values were imputed successfully.

## 6. Use different algorithms that support missing values.

Not all algorithms fail when there is missing data. There are algorithms that can be made robust to missing data, such as k-Nearest Neighbors that can ignore a column from a distance measure when a value is missing.

There are also algorithms that can use the missing value as a unique and different value when building the predictive model, such as classification and regression trees.

I hope this has helped you in some way.